Cache doesn't find `getMetadata` method although it is present

The `Cache` implementation doesn't find the `getMetadata` method although it is present:

~~~
2019-12-10T14:32:21.596Z ERROR [ReflectorRunnable] Failure in list-watch: java.lang.ReflectiveOperationException: java.lang.NoSuchMethodException: io.enmasse.iot.model.v1.IoTProject.getMetadata()
java.lang.RuntimeException: java.lang.ReflectiveOperationException: java.lang.NoSuchMethodException: io.enmasse.iot.model.v1.IoTProject.getMetadata()
	at io.fabric8.kubernetes.client.informers.cache.Cache.metaNamespaceKeyFunc(Cache.java:365)
	at io.fabric8.kubernetes.client.informers.cache.DeltaFIFO.keyOf(DeltaFIFO.java:381)
	at io.fabric8.kubernetes.client.informers.cache.DeltaFIFO.replace(DeltaFIFO.java:165)
	at io.fabric8.kubernetes.client.informers.cache.ReflectorRunnable.syncWith(ReflectorRunnable.java:145)
	at io.fabric8.kubernetes.client.informers.cache.ReflectorRunnable.run(ReflectorRunnable.java:69)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ReflectiveOperationException: java.lang.NoSuchMethodException: io.enmasse.iot.model.v1.IoTProject.getMetadata()
	at io.fabric8.kubernetes.client.utils.ReflectUtils.objectMetadata(ReflectUtils.java:31)
	at io.fabric8.kubernetes.client.informers.cache.Cache.metaNamespaceKeyFunc(Cache.java:355)
	... 10 more
Caused by: java.lang.NoSuchMethodException: io.enmasse.iot.model.v1.IoTProject.getMetadata()
	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2475)
	at io.fabric8.kubernetes.client.utils.ReflectUtils.objectMetadata(ReflectUtils.java:28)
	... 11 more
~~~

This is due to the fact that `ReflectUtils` uses `getDeclaredMethod`, which will only find methods which are directly declared on the class:

https://github.com/fabric8io/kubernetes-client/blob/e2a8c9f7b9329099b1ef88e3f1d27f7a4056ab71/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/ReflectUtils.java#L28-L29

Instead of doing reflection, the code should check for `instanceof HasMetadata` and maybe then use `Class#getMethod`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 ### 4.7-SNAPSHOT
 #### Bugs
 * Fix #1850: Add option to disable timestamps in build logs on Openshift
+* Fix #1902: Fix the usage of reflection, so that `getMetadata` is detected properly
 
 #### Improvements
 * Fix #1880: Remove use of reapers manually doing cascade deletion of resources, leave it upto Kubernetes APIServer

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/ReflectUtils.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/ReflectUtils.java
@@ -15,6 +15,9 @@
  */
 package io.fabric8.kubernetes.client.utils;
 
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.KubernetesResourceList;
+import io.fabric8.kubernetes.api.model.ListMeta;
 import io.fabric8.kubernetes.api.model.ObjectMeta;
 
 import java.lang.reflect.InvocationTargetException;
@@ -24,8 +27,11 @@ import java.util.List;
 public class ReflectUtils {
 
   public static ObjectMeta objectMetadata(Object obj) throws ReflectiveOperationException {
+    if (obj instanceof HasMetadata) {
+      return ((HasMetadata) obj).getMetadata();
+    }
     try {
-      Method mdField = obj.getClass().getDeclaredMethod("getMetadata");
+      Method mdField = obj.getClass().getMethod("getMetadata");
       return (ObjectMeta) mdField.invoke(obj);
     } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
       throw new ReflectiveOperationException(e);
@@ -36,16 +42,22 @@ public class ReflectUtils {
     return objectMetadata(obj).getNamespace();
   }
 
-  public static ObjectMeta listMetadata(Object listObj) throws ReflectiveOperationException {
+  public static ListMeta listMetadata(Object listObj) throws ReflectiveOperationException {
+    if (listObj instanceof KubernetesResourceList) {
+      return ((KubernetesResourceList<?>) listObj).getMetadata();
+    }
     try {
-      Method mdField = listObj.getClass().getDeclaredMethod("getMetadata");
-      return (ObjectMeta) mdField.invoke(listObj);
+      Method mdField = listObj.getClass().getMethod("getMetadata");
+      return (ListMeta) mdField.invoke(listObj);
     } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
       throw new ReflectiveOperationException(e);
     }
   }
 
   public static <ApiType> List<ApiType> getItems(Object listObj) throws ReflectiveOperationException {
+    if (listObj instanceof KubernetesResourceList) {
+      return ((KubernetesResourceList) listObj).getItems();
+    }
     try {
       Method getItemsMethod = listObj.getClass().getMethod("getItems");
       return (List<ApiType>) getItemsMethod.invoke(listObj);

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/ReflectUtilsTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/ReflectUtilsTest.java
@@ -1,0 +1,96 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.kubernetes.client.utils;
+
+import static org.junit.jupiter.api.Assertions.assertSame;
+
+import org.junit.jupiter.api.Test;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+
+public class ReflectUtilsTest {
+
+  static class Foo {
+    private ObjectMeta metadata;
+    public void setMetadata(ObjectMeta metadata) {
+      this.metadata = metadata;
+    }
+    public ObjectMeta getMetadata() {
+      return metadata;
+    }
+  }
+
+  static class Bar extends Foo {
+  }
+
+  @SuppressWarnings("serial")
+  static abstract class AbstractBaz implements HasMetadata {
+    private ObjectMeta metadata;
+    private String apiVersion;
+
+    @Override
+    public void setMetadata(ObjectMeta metadata) {
+      this.metadata = metadata;
+    }
+    @Override
+    public ObjectMeta getMetadata() {
+      return this.metadata;
+    }
+    @Override
+    public void setApiVersion(String version) {
+     this.apiVersion = version;
+    }
+    @Override
+    public String getApiVersion() {
+      return this.apiVersion;
+    }
+    
+  }
+
+  @SuppressWarnings("serial")
+  static class Baz extends AbstractBaz {
+    @Override
+    public String getKind() {
+      return Baz.class.getSimpleName();
+    }
+  }
+
+  @Test
+  public void testDirectReflection () throws ReflectiveOperationException {
+    ObjectMeta meta = new ObjectMeta();
+    Foo foo = new Foo();
+    foo.setMetadata(meta);
+    assertSame(meta, ReflectUtils.objectMetadata(foo));
+  }
+
+  @Test
+  public void testDerivedReflection () throws ReflectiveOperationException {
+    ObjectMeta meta = new ObjectMeta();
+    Foo bar = new Bar();
+    bar.setMetadata(meta);
+    assertSame(meta, ReflectUtils.objectMetadata(bar));
+  }
+
+  @Test
+  public void testInstanceOf() throws ReflectiveOperationException {
+    ObjectMeta meta = new ObjectMeta();
+    Baz baz = new Baz();
+    baz.setMetadata(meta);
+    assertSame(meta, ReflectUtils.objectMetadata(baz));
+  }
+
+}

